renku/core/management/migrations/m_0005__2_cwl.py

```diff
@@ -0,0 +1,361 @@
+    commit = client.find_previous_commit(path)
+    run = Run()
+
+    name = '{0}_migrated.yaml'.format(uuid.uuid4().hex)
+
+    run.path = (client.workflow_path / name).relative_to(client.path)
+
+    for step in workflow.steps:
+        path = client.workflow_path / step.run
+        subrun = parse_cwl_cached(str(path))
+
+        subprocess, _ = _migrate_single_step(
+            client, subrun, path, commit=commit
+        )
+        subprocess.path = run.path
+        run.add_subprocess(subprocess)
+
+    with with_reference(run.path):
+        wf = WorkflowRun.from_run(run, client, run.path, commit=commit)
+        wf.to_yaml()
+        client.add_to_activity_index(wf)
+
+    return wf, run.path
+
+
+def _entity_from_path(client, path, commit):
+    """Gets the entity associated with a path."""
+    client, commit, path = client.resolve_in_submodules(
+        client.find_previous_commit(path, revision=commit.hexsha),
+        path,
+    )
+
+    entity_cls = Entity
+    if (client.path / path).is_dir():
+        entity_cls = Collection
+
+    if str(path).startswith(os.path.join(client.renku_home, client.DATASETS)):
+        return client.load_dataset_from_path(path, commit=commit)
+    else:
+        return entity_cls(
+            commit=commit,
+            client=client,
+            path=str(path),
+        )
+
+
+def _invalidations_from_commit(client, commit):
+    """Gets invalidated files from a commit."""
+    results = []
+    collections = dict()
+    for file_ in commit.diff(commit.parents or NULL_TREE):
+        # only process deleted files (note they appear as ADDED)
+        # in this backwards diff
+        if file_.change_type != 'A':
+            continue
+        path_ = Path(file_.a_path)
+        entity = _get_activity_entity(
+            client, commit, path_, collections, deleted=True
+        )
+
+        results.append(entity)
+
+    return results
+
+
+def _get_activity_entity(client, commit, path, collections, deleted=False):
+    """Gets the entity associated with this Activity and path."""
+    client, commit, path = client.resolve_in_submodules(
+        commit,
+        path,
+    )
+    output_path = client.path / path
+    parents = list(output_path.relative_to(client.path).parents)
+
+    collection = None
+    members = []
+    for parent in reversed(parents[:-1]):
+        if str(parent) in collections:
+            collection = collections[str(parent)]
+        else:
+            collection = Collection(
+                client=client,
+                commit=commit,
+                path=str(parent),
+                members=[],
+                parent=collection,
+            )
+            members.append(collection)
+            collections[str(parent)] = collection
+
+        members = collection.members
+
+    entity_cls = Entity
+    if (client.path / path).is_dir():
+        entity_cls = Collection
+
+    # TODO: use a factory method to generate the entity
```

Comment:
Can we open a ticket for this and put the ticket reference here as well?

Reply:
Oh that's because of copy-pasting from here: https://github.com/SwissDataScienceCenter/renku-python/blob/master/renku/core/models/provenance/activities.py#L160 (since migrations should be self-contained and not depend on code that might change).

I'm not sure if we want a factory method there.